packages/core/js-client/src/compilerSupport/conversions.ts

```diff
@@ -14,222 +14,153 @@
- * @returns value represented in typescript
- */
-export const aqua2ts = (value: JSONValue, type: NonArrowType): JSONValue => {
-  const res = match(type)
-    .with({ tag: "nil" }, () => {
-      return null;
-    })
-    .with({ tag: "option" }, (opt) => {
-      assert(Array.isArray(value), "Should not be possible, bad types");
-
-      if (value.length === 0) {
-        return null;
-      } else {
-        return aqua2ts(value[0], opt.type);
-      }
-    })
-    .with({ tag: "scalar" }, { tag: "bottomType" }, { tag: "topType" }, () => {
-      return value;
-    })
-    .with({ tag: "array" }, (arr) => {
-      assert(Array.isArray(value), "Should not be possible, bad types");
-      return value.map((y) => {
-        return aqua2ts(y, arr.type);
-      });
-    })
-    .with({ tag: "struct" }, (x) => {
-      return Object.entries(x.fields).reduce((agg, [key, type]) => {
-        const val = aqua2ts(value[key], type);
-        return { ...agg, [key]: val };
-      }, {});
-    })
-    .with({ tag: "labeledProduct" }, (x) => {
-      return Object.entries(x.fields).reduce((agg, [key, type]) => {
-        const val = aqua2ts(value[key], type);
-        return { ...agg, [key]: val };
-      }, {});
-    })
-    .with({ tag: "unlabeledProduct" }, (x) => {
-      return x.items.map((type, index) => {
-        return aqua2ts(value[index], type);
-      });
-    })
-    // uncomment to check that every pattern in matched
-    // .exhaustive();
-    .otherwise(() => {
-      throw new Error("Unexpected tag: " + jsonify(type));
-    });
 
-  return res;
-};
-
-/**
- * Convert call service arguments list from their aqua representation to representation in typescript
- * @param req - call service data
- * @param arrow - aqua type definition
- * @returns arguments in typescript representation
- */
-export const aquaArgs2Ts = (
-  req: CallServiceData,
-  arrow: ArrowWithoutCallbacks,
-): JSONArray => {
-  const argTypes = match(arrow.domain)
-    .with({ tag: "labeledProduct" }, (x) => {
-      return Object.values(x.fields);
-    })
-    .with({ tag: "unlabeledProduct" }, (x) => {
-      return x.items;
-    })
-    .with({ tag: "nil" }, (x) => {
-      return [];
-    })
-    // uncomment to check that every pattern in matched
-    // .exhaustive()
-    .otherwise(() => {
-      throw new Error("Unexpected tag: " + jsonify(arrow.domain));
-    });
+import { ParticleContext } from "../jsServiceHost/interfaces.js";
 
-  if (req.args.length !== argTypes.length) {
-    throw new Error(
-      `incorrect number of arguments, expected: ${argTypes.length}, got: ${req.args.length}`,
-    );
-  }
+import { MaybePromise } from "./types.js";
 
-  return req.args.map((arg, index) => {
-    return aqua2ts(arg, argTypes[index]);
-  });
-};
+export function aqua2ts(
+  value: JSONValue,
+  schema: NonArrowSimpleType,
+): JSONValue {
+  if (schema.tag === "nil") {
+    return null;
+  } else if (schema.tag === "option") {
+    if (!Array.isArray(value)) {
```

Comment:
let's move everything under this condition to a separate function? like `validateOption` or something. Gotta reduce LOCs of `aqua2ts`